File: Week_04/102/binary_tree_level_order_traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	res := make([][]int, 0)
	if root == nil {
		return res
	}

	queue := []*TreeNode{root}
	level := 0
	for len(queue) > 0 {
		n := len(queue)
		for i := 0; i < n; i++ {
			if len(res) == level {
				res = append(res, []int{queue[i].Val})
			} else {
				res[level] = append(res[level], queue[i].Val)
			}
			if queue[i].Left != nil {
				queue = append(queue, queue[i].Left)
			}
			if queue[i].Right != nil {
				queue = append(queue, queue[i].Right)
			}
		}
		queue = queue[n:]
		level++
	}
	return res
}

File: Week_04/102/binary_tree_level_order_traversal.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        res = []

        if root is None:
            return res

        queue = [root]
        level = 0
        while queue:
            n = len(queue)
            for i in range(n):
                if level == len(res):
                    res.append([queue[i].val])
                else:
                    res[level].append(queue[i].val)
                if queue[i].left is not None:
                    queue.append(queue[i].left)
                if queue[i].right is not None:
                    queue.append(queue[i].right)
            level += 1
            queue = queue[n:]
        return res
